swift-sdk/Internal/InAppManager.swift

```diff
@@ -84,75 +168,119 @@ class InAppManager : IterableInAppManagerProtocol {
+        guard waitTimeInterval <= 0 else {
+            ITBInfo("Need to wait for: \(waitTimeInterval)")
+            queue.asyncAfter(deadline: .now() + waitTimeInterval) {
+                self.processMessages()
             }
-        } else if newMessages.count > 1 {
-            if let message = inAppDelegate.onNew(batch: newMessages) {
-                // found content to show
-                self.show(message: message, consume: true)
-                
-                newMessages.filter { $0.messageId != message.messageId }.forEach { markAsSkipped(message: $0) }
+            return
+        }
+        
+        // We can only check applicationState from main queue so need to do the following
+        DispatchQueue.main.async {
+            if self.applicationStateProvider.applicationState == .active {
+                self.queue.async {
+                    self.processOneMessage()
+                }
             } else {
-                ITBInfo("skipping inApp batch")
-                newMessages.forEach {markAsSkipped(message: $0)}
+                ITBInfo("Application is not active")
             }
         }
     }
     
-    private func markAsSkipped(message: IterableInAppMessage) {
-        if let message = messagesMap[message.messageId] {
-            message.skipped = true
-            messagesMap.updateValue(message, forKey: message.messageId)
+    // How long do we have to wait before showing the message
+    // > 0 means wait, otherwise we are good to show
+    private func getWaitTimeInterval() -> Double {
+        if let lastDismissedTime = lastDismissedTime {
+            let nextShowingTime = Date(timeInterval: retryInterval, since: lastDismissedTime)
+            if dateProvider.currentDate >= nextShowingTime {
+                return 0.0
+            } else {
+                return nextShowingTime.timeIntervalSinceReferenceDate - self.dateProvider.currentDate.timeIntervalSinceReferenceDate
+            }
         } else {
-            // Should never happen
-            ITBError("Did not find message")
+            // we have not shown any messages
+            return 0.0
         }
     }
+    
+    // go through one loop of client side messages, and show the first that is not processed
+    private func processOneMessage() {
+        ITBDebug()
 
-    // Adds new messages to map and returns the new messages
-    private func mergeAndGetNewMessages(messages: [IterableInAppMessage]) -> [IterableInAppMessage] {
-        return messages.reduce(into: [IterableInAppMessage]()) { (result, message) in
-            if !messagesMap.contains(where: { $0.key == message.messageId}) {
-                messagesMap[message.messageId] = message
-                result.append(message)
+        for message in messagesMap.values.filter({ $0.processed == false }) {
+            ITBDebug("campaignId: \(message.campaignId)")
+            message.processed = true
+            if inAppDelegate.onNew(message: message) == .show {
```

Comment:
Since we're going all the way to `showInApp` before we know if it can be shown, it may be possible that the app will receive a new in-app message while another is on the screen, call `onNew` and then fail to display the message in the displayer. Is there a way to add a check before that?

Reply:
If showInApp fails for newly arriving message due to a message already being shown then it will retry after retryInterval.

Reply:
Basically showOneMessage keeps retrying until it succeeds. Hmm... although I can see a problem with this if at the end of retryInterval the app goes to background.

Reply:
Oh yeah you're right. Then the issue is that if the app goes to background within 30s, we'll never show this message because it was marked as processed but not shown. Even if the app doesn't go to background, the in-app message may be shown at some later point, and the app won't know when. onNew should probably be called right before the SDK is going to show the message (and can do that) so that the app can decide if now is a good time to show it.

Reply:
Makes sense. For that we will need to add a method to check for 'canShowMessage()' method on the InAppDisplayer. I will do that.